crates/wasi-common/src/snapshots/preview_1.rs

```diff
@@ -42,6 +42,45 @@ impl types::UserErrorConversion for WasiCtx {
     }
 }
 
+fn is_error_notdir(e: &Error) -> bool {
+    if e.is::<ErrorKind>() {
+        let e = e.downcast_ref::<ErrorKind>().unwrap();
+        if let ErrorKind::Notdir = *e {
+            true
+        } else {
+            false
+        }
+    } else if e.is::<std::io::Error>() {
+        let e = e.downcast_ref::<std::io::Error>().unwrap();
+        #[cfg(unix)]
+        {
+            e.raw_os_error() == Some(rustix::io::Errno::NOTDIR.raw_os_error())
+        }
+        #[cfg(windows)]
+        {
+            e.raw_os_error() == Some(windows_sys::Win32::Foundation::ERROR_DIRECTORY as _)
+        }
+    } else {
+        false
+    }
+}
+
+fn is_error_noent(e: &Error) -> bool {
+    if e.is::<ErrorKind>() {
+        let e = e.downcast_ref::<ErrorKind>().unwrap();
+        if let ErrorKind::Noent = *e {
+            true
+        } else {
+            false
+        }
+    } else if e.is::<std::io::Error>() {
+        let e = e.downcast_ref::<std::io::Error>().unwrap();
+        e.kind() == std::io::ErrorKind::NotFound
+    } else {
+        false
+    }
+}
```

Comment:
It looks like the error handling in this crate can benefit from quite some refactoring as this does not look like a maintainable way to match on errors.

Note, that this gets much worse for external users of this crate providing e.g. multiple `WasiFile` or `WasiDir` implementations, since users must return `wasi-common::Error` constructed internally from an `ErrorKind` in custom implementations or a few other special cases, e.g. from https://github.com/bytecodealliance/wasmtime/blob/29c7de73401347af6d8c7111bdb7a11f1b04c3b7/crates/wasi-common/src/snapshots/preview_1.rs#L48-L64, anything else results in a trap and terminates execution.
Every assertion on the error kind now requires a downcast and this runtime reflection step can fail if e.g. the `wasi-common` crate changes the set of specific error types returned *or* downcasted internally on update.

So perhaps these utilities should actually be public to expose them to users who may need to match on the error kind from outside this crate?

Reply:
I agree that the current code is not great. It sounds like another way to approach this would be to get rid of the `anyhow` and downcasting altogether and just use an enum. I may be up for doing the work for this refactoring, but in this PR I'd like to focus on just fixing the underlying bug first, and we can figure out refactoring separately.

Reply:
Yes, I agree that does not belong in this PR :)
Another option could be an associated `type Error: AsRef<ErrorKind>` or `type Error: AsRef<Errno>`, whichever is the most applicable.